File: src/JavaScript/Decimal-to-Factorial-and-Back.js

const naturalSequence = require("./toolkit/Vector").naturalSequence;
const factorial = require("./toolkit/Arithmetic").factorial;
const hadamardProduct = require("./toolkit/Matrix").hadamardProduct;
const sum = require("./toolkit/Statistics").sum;


// 字符串本身就是 iterator, 可以用索引 [] 访问
const code = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const encode = new Map, decode = new Map;
for (let i = 0; i <= 35; i++) {
  encode.set(i, code[i]);
  decode.set(code[i], i);
}


function dec2FactString(nb) {
  // 寻找最大的基
  let maxBase = 1;
  while (factorial(maxBase + 1) <= nb) {
    maxBase++;
  }

  // 对应这组基的系数向量
  const factors = [];
  for (let i = maxBase; i >= 0; i--) {
    char = encode.get(Math.floor(nb / factorial(i)));
    factors.push(char);
    nb %= factorial(i);
  }

  return factors.join('');
}

console.log(dec2FactString(463));


function factString2Dec(string) {
  const factors = string.split('').reverse().map(char => decode.get(char));
  const bases = naturalSequence(string.length, start = 0).map(factorial);
  return sum(hadamardProduct(factors, bases));
}

console.log(factString2Dec("341010"));

File: src/Julia/Decimal-to-Factorial-and-Back.jl
## Decimal-to-Factorial-and-Back.jl

using Pipe, Test


const code = ['0':'9'; 'A':'Z']
const encode = Dict(i - 1 => code[i] for i = 1:length(code))
const decode = Dict(code[i] => i - 1 for i = 1:length(code))


"""
将整数线性分解为一组基，返回系数连接成的字符串
"""
function dec2FactString(nb::Int)::String
    # 先找最大的基
    max_base = 1
    while factorial(max_base + 1) <= nb
        max_base += 1
    end

    # 从大到小，确定每个基对应的系数
    factors = ""
    for base ∈ max_base:-1:0
        factors *= encode[nb÷factorial(base)] # 这里的*表示字符串连接
        nb %= factorial(base)
    end

    factors
end

@test dec2FactString(463) == "341010"



"""
已知基（阶乘）和系数连接成的字符串，计算整数
"""
function factString2Dec(str::String)::Int
    # 逆序的系数向量
    factors = @pipe collect(str) .|> decode[_] |> reverse

    # 基向量，从 0! 到 max_base!
    bases = 0:(length(factors)-1) .|> factorial

    sum(factors .* bases)
end

@test factString2Dec("341010") == 463